File: goldencheck/engine/confidence.py
"""Post-scan confidence processing."""
from __future__ import annotations
from collections import defaultdict
from dataclasses import replace
from goldencheck.models.finding import Finding, Severity

__all__ = ["apply_confidence_downgrade", "apply_corroboration_boost"]


def apply_corroboration_boost(findings: list[Finding]) -> list[Finding]:
    """Boost confidence for columns flagged by multiple profilers.

    - 2 distinct WARNING/ERROR checks on the same column: +0.1
    - 3+ distinct WARNING/ERROR checks on the same column: +0.2 (exclusive tiers)
    - Capped at 1.0
    - Uses dataclasses.replace() — original findings are never mutated.
    """
    # Count distinct WARNING/ERROR checks per column
    checks_per_col: dict[str, set[str]] = defaultdict(set)
    for f in findings:
        if f.severity in (Severity.ERROR, Severity.WARNING):
            checks_per_col[f.column].add(f.check)

    result = []
    for f in findings:
        col_count = len(checks_per_col.get(f.column, set()))
        if col_count >= 3:
            boost = 0.2
        elif col_count >= 2:
            boost = 0.1
        else:
            boost = 0.0

        if boost > 0 and f.severity in (Severity.ERROR, Severity.WARNING):
            new_conf = min(f.confidence + boost, 1.0)
            result.append(replace(f, confidence=new_conf))
        else:
            result.append(f)
    return result


def apply_confidence_downgrade(findings: list[Finding], llm_boost: bool) -> list[Finding]:
    """Downgrade low-confidence findings to INFO when LLM boost is not enabled.

    If llm_boost=True, return findings unchanged (LLM will verify them).
    If llm_boost=False, any WARNING/ERROR with confidence < 0.5 is downgraded to INFO
    with a suffix explaining how to get better accuracy.
    """
    if llm_boost:
        return list(findings)

    result = []
    for f in findings:
        if f.confidence < 0.5 and f.severity in (Severity.ERROR, Severity.WARNING):
            result.append(replace(
                f,
                severity=Severity.INFO,
                message=f"{f.message} (low confidence — use --llm-boost to verify)",
            ))
        else:
            result.append(f)
    return result

File: packages/goldencheck-js/src/core/engine/confidence.ts
/**
 * Post-scan confidence processing.
 * Port of goldencheck/engine/confidence.py.
 */

import type { Finding } from "../types.js";
import { Severity, replaceFinding } from "../types.js";

/**
 * Boost confidence for columns flagged by multiple profilers.
 * - 2 distinct WARNING/ERROR checks on same column: +0.1
 * - 3+ distinct checks: +0.2 (exclusive tiers)
 * - Capped at 1.0
 * - Returns new array — originals never mutated.
 */
export function applyCorroborationBoost(findings: readonly Finding[]): Finding[] {
  // Count distinct WARNING/ERROR checks per column
  const checksPerCol = new Map<string, Set<string>>();
  for (const f of findings) {
    if (f.severity === Severity.ERROR || f.severity === Severity.WARNING) {
      let checks = checksPerCol.get(f.column);
      if (!checks) {
        checks = new Set();
        checksPerCol.set(f.column, checks);
      }
      checks.add(f.check);
    }
  }

  return findings.map((f) => {
    const colCount = checksPerCol.get(f.column)?.size ?? 0;
    let boost: number;
    if (colCount >= 3) boost = 0.2;
    else if (colCount >= 2) boost = 0.1;
    else boost = 0.0;

    if (boost > 0 && (f.severity === Severity.ERROR || f.severity === Severity.WARNING)) {
      return replaceFinding(f, { confidence: Math.min(f.confidence + boost, 1.0) });
    }
    return f;
  });
}

/**
 * Downgrade low-confidence findings to INFO when LLM boost is not enabled.
 * If llmBoost=true, returns findings unchanged.
 * If llmBoost=false, WARNING/ERROR with confidence < 0.5 → INFO.
 */
export function applyConfidenceDowngrade(findings: readonly Finding[], llmBoost: boolean): Finding[] {
  if (llmBoost) return [...findings];

  return findings.map((f) => {
    if (f.confidence < 0.5 && (f.severity === Severity.ERROR || f.severity === Severity.WARNING)) {
      return replaceFinding(f, {
        severity: Severity.INFO,
        message: `${f.message} (low confidence — use --llm-boost to verify)`,
      });
    }
    return f;
  });
}
